holodeck/cyutils.pyx

```diff
@@ -896,6 +897,667 @@ cdef double[:, :] _sam_poisson_gwb(
     return gwb
 
 
+def ss_bg_hc(number, h2fdf, nreals, normal_threshold=1e10):
+    """ Calculates the characteristic strain from loud single sources and a background of all other sources.
+
+    Parameters
+    ----------
+    number : [M, Q, Z, F] ndarray
+        number in each bin
+    h2fdf : [M, Q, Z, F] ndarray
+        strain squared x frequency / frequency bin width for each bin
+    nreals
+        number of realizations
+
+    Returns
+    -------
+    hc2ss : (F, R) Ndarray of scalars
+    hc2bg : (F, R) Ndarray of scalars
+    ssidx : (3, F, R) Ndarray of ints
+        Index of the loudest single source, -1 if there are none at the frequency/realization.
+
+    """
+    
+    cdef long[:] shape = np.array(number.shape)
+    F = shape[3]
+    R = nreals
+    cdef np.ndarray[np.double_t, ndim=2] hc2ss = np.zeros((F,R))
+    cdef np.ndarray[np.double_t, ndim=2] hc2bg = np.zeros((F,R))
+    cdef np.ndarray[np.long_t, ndim=3] ssidx = np.zeros((3,F,R), dtype=int)
+    _ss_bg_hc(shape, h2fdf, number, nreals, normal_threshold, 
+                hc2ss, hc2bg, ssidx)
+    # print(hc2ss, hc2bg, ssidx)
+    return hc2ss, hc2bg, ssidx
+    
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+@cython.cdivision(True)
+cdef void _ss_bg_hc(long[:] shape, double[:,:,:,:] h2fdf, double[:,:,:,:] number,
+            long nreals, long thresh,
+            double[:,:] hc2ss, double[:,:] hc2bg, long[:,:,:] ssidx):
+    """
+    Calculates the characteristic strain from loud single sources and a background of all other sources.
```

Comment:
This is just the single loudest source, right?

Reply:
yep. It's used by ss_by_cdefs, but there's no reason to ever use this in production. It returns hc_ss as an (F,R) 2Darray, but it's better to use ss_gws(...,loudest=1) to return an (F,R,1) 3Darray even if you only want the single loudest sources, because that's the shape everything else (e.g. plotting functions, library generation) are designed to use. I think the cyutils function that ss_gws() uses is probably just better in general. Only time to use ss_by_cdefs could be if you want ssidx (indices of single sources).